crypto/evp/e_aes.c

```diff
@@ -1791,6 +1794,15 @@ static int s390x_aes_gcm_tls_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
     if (out != in || len < (EVP_GCM_TLS_EXPLICIT_IV_LEN + EVP_GCM_TLS_TAG_LEN))
         return -1;
 
+    /*
+     * Check for too many keys as per FIPS 140-2 IG A.5 "Key/IV Pair Uniqueness
+     * Requirements from SP 800-38D".  The requirements is for one party to the
+     * communication to fail after 2^64 - 1 keys.  We do this on the encrypting
+     * side only.
+     */
+    if (ctx->encrypt && ++gctx->tls_enc_records == 0)
```

Comment:
should this set an error code?

Reply:
Given that the test cannot ever succeed, there doesn't seem to be much point.  The TLS stack will catch this earlier on.  We don't want the TLS stack inside the FIPS boundary but we need a check inside.

If there is support for this, I'll add an error code.

Reply:
Can you guarantee that only libssl will use this partical EVP?

Reply:
The only way to execute this code is by making a _ctrl_ call with the control argument `EVP_CTRL_AEAD_TLS1_AAD`.  While it would theoretically be possible to make the _ctrl_ call outside of TLS it seems unlikely.

This is a FIPS mandated check that doesn't connect well with reality.  This check will only kick in after 2<sup>64</sup>-1 _records_ are processed in a _single_ session.  That is an impressive amount of data -- over a decade nonstop at 1,000 GBytes/s.